src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,63 +807,206 @@ ErrorIfInsertPartitionColumnDoesNotMatchSelect(Query *query, RangeTblEntry *inse
+							exprTypeStr = "a function call";
+						}
+						break;
+					}
+				}
+				break;
+
+				case T_Aggref:
+				{
+					exprTypeStr = "an aggregation";
+				}
+				break;
+
+				case T_CaseExpr:
+				{
+					exprTypeStr = "a case expression";
+				}
+				break;
+
+				case T_CoalesceExpr:
+				{
+					exprTypeStr = "a coalesce expression";
+				}
+				break;
+
+				case T_RowExpr:
+				{
+					exprTypeStr = "a row expression";
+				}
+				break;
+
+				case T_MinMaxExpr:
+				{
+					exprTypeStr = "a min/max expression";
+				}
+
+				break;
+
+				case T_CoerceViaIO:
+				{
+					exprTypeStr = "an explicit coercion";
+				}
+
+				break;
+
+				default:
+				{
+					exprTypeStr = exprTypeDefault;
+				}
 				break;
 			}
 
-			partitionColumnsMatch = true;
-			*selectPartitionColumnTableId = subqeryTargetEntry->resorigtbl;
+			errorDetail = psprintf(errorDetailTemplate, exprTypeStr);
 
-			break;
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO .. SELECT query "
+								   "due to partition column restrictions"),
+							errdetail("%s", errorDetail),
+							errhint("Ensure that target partition column matches with a "
+									"plain partition column of a distributed table in "
+									"the subquery.")));
+		}
+
+		/*
+		 * Insert target expression could only be non-var if the select target
+		 * entry does not have the same type (i.e., target column requires casting).
+		 */
+		if (!IsA(targetEntry->expr, Var))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO .. SELECT query "
+								   "due to data type mismatch of partition columns"),
+							errdetail(
+								"The data type of the target table's partition column "
+								"should exactly match the data type of the "
+								"corresponding target list entry in the subquery.")));
 		}
+
+		/* finally, check that the select target column is a partition column */
+		if (!IsPartitionColumnRecursive(subqeryTargetEntry->expr, subquery))
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot plan distributed INSERT INTO .. SELECT query "
+								   "due to mismatch of partition columns"),
+							errdetail(
+								"The target table's partition column "
+								"should match a partition column on the "
+								"corresponding target list entry of the subquery.")));
+		}
+
+		/* set this for the caller function */
+		*selectPartitionColumnTableId = subqeryTargetEntry->resorigtbl;
+
+		break;
 	}
 
-	if (!partitionColumnsMatch)
+
+	if (!targetTableHasPartitionColumn)
 	{
 		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-						errmsg("SELECT query should return bare partition column on "
-							   "the same ordinal position as the INSERT's partition "
-							   "column")));
+						errmsg("cannot plan distributed INSERT INTO .. SELECT query"),
+						errdetail("The query should contain the partition column of "
```

Comment:
_INSERT INTO ... SELECT queries must include the target table's partition column_?